File: src/MemoryKVLib_GTest/BoundaryTests.cpp

#include "pch.h"
#include <gtest/gtest.h>
#include "../MemoryKVLib/MemoryKV.h"
#include <string>
#include <vector>

#include "MockLogger.h"

class BoundaryTest : public ::testing::Test {
protected:
    void SetUp() override {
        kv = new MemoryKV(L"test_client", std::make_unique<MockLogger>(true));
        std::cout << "setup\n";
    }

    void TearDown() override {
        delete kv;
    }

    MemoryKV* kv;
};

// 测试最大键长度
TEST_F(BoundaryTest, MaxKeyLength) {
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    kv->Open(L"MaxKeyLength", options);
    
    // 测试最大长度键
    std::wstring maxKey(64-1, L'a');
    std::wstring value = L"test_value";
    EXPECT_TRUE(kv->Put(maxKey, value));
    
    const wchar_t* result = kv->Get(maxKey);
    EXPECT_STREQ(result, value.c_str());
    
    // 测试超长键
    std::wstring tooLongKey(64, L'a');
    EXPECT_FALSE(kv->Put(tooLongKey, value));
}

// 测试最大值长度
TEST_F(BoundaryTest, MaxValueLength) {
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    kv->Open(L"MaxValueLength", options);
    
    // 测试最大长度值
    std::wstring key = L"test_key";
    std::wstring maxValue(256-1, L'a');
    EXPECT_TRUE(kv->Put(key, maxValue));
    
    const wchar_t* result = kv->Get(key);
    EXPECT_STREQ(result, maxValue.c_str());
    
    // 测试超长值
    std::wstring tooLongValue(256, L'a');
    EXPECT_FALSE(kv->Put(key, tooLongValue));
}

// 测试最大并发连接数
TEST_F(BoundaryTest, MaxConcurrentConnections) {
    const int maxConnections = 100;
    std::vector<MemoryKV*> clients;
    
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    // 创建最大数量的连接
    for (int i = 0; i < maxConnections; ++i) {
        auto client = new MemoryKV(L"test_client", std::make_unique<MockLogger>());
        client->Open(L"MaxConcurrentConnections", options);
        clients.push_back(client);
    }
    
    // 尝试创建额外连接
    auto extraClient = new MemoryKV(L"test_client", std::make_unique<MockLogger>());
    extraClient->Open(L"MaxConcurrentConnections", options);
    delete extraClient;
    
    // 清理
    for (auto client : clients) {
        delete client;
    }
}

// 测试内存使用限制
TEST_F(BoundaryTest, MemoryUsageLimit) {
    const int maxMemoryMB = 100; // 100KB 内存限制
    const int valueSize = 256; // 1KB 值大小
    
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = valueSize;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1; //do not allow expansion
    options.LogLevel = 0;
    
    kv->Open(L"MemoryUsageLimit", options);
    
    // 尝试写入超过内存限制的数据
    std::wstring largeValue(valueSize-1, L'a');
    for (int i = 0; i < maxMemoryMB + 1; ++i) {
        std::wstring key = L"key_" + std::to_wstring(i);
        std::cout << i << "\n";
        if (i < maxMemoryMB) {
            EXPECT_TRUE(kv->Put(key, largeValue));
        } else {
            EXPECT_FALSE(kv->Put(key, largeValue));
        }
    }
}

// 测试空键和空值
TEST_F(BoundaryTest, EmptyKeyAndValue) {
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    kv->Open(L"EmptyKeyAndValue", options);
    
    // 测试空键
    EXPECT_FALSE(kv->Put(L"", L"value"));
    
    // 测试空值
    EXPECT_TRUE(kv->Put(L"key", L""));
    const wchar_t* result = kv->Get(L"key");
    EXPECT_STREQ(result, L"");
    kv->Remove(L"key");
}

// 测试特殊字符
TEST_F(BoundaryTest, SpecialCharacters) {
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    kv->Open(L"SpecialCharacters", options);
    
    // 测试特殊字符键
    std::wstring specialKey = L"key!@#$%^&*()_+-=[]{}|;:'\",.<>?/\\";
    std::wstring value = L"test_value";
    EXPECT_TRUE(kv->Put(specialKey, value));
    
    const wchar_t* result = kv->Get(specialKey);
    EXPECT_STREQ(result, value.c_str());
    kv->Remove(specialKey);
    
    // 测试特殊字符值
    std::wstring key = L"test_key";
    std::wstring specialValue = L"value!@#$%^&*()_+-=[]{}|;:'\",.<>?/\\";
    EXPECT_TRUE(kv->Put(key, specialValue));
    
    result = kv->Get(key);
    EXPECT_STREQ(result, specialValue.c_str());
    kv->Remove(key);
}

// 测试缓冲区大小
TEST_F(BoundaryTest, BufferSize) {
    ConfigOptions options;
    options.MaxKeySize = 64;
    options.MaxValueSize = 256;
    options.MaxBlocksPerMmf = 100;
    options.MaxMmfCount = 1;
    options.LogLevel = 0;
    
    kv->Open(L"BufferSize", options);
    
    // 测试不同大小的键和值
    for (int i = 1; i <= 64-1; ++i) {
        std::wstring key(i, L'a');
        std::wstring value(i, L'b');
        EXPECT_TRUE(kv->Put(key, value));
        
        const wchar_t* result = kv->Get(key);
        EXPECT_STREQ(result, value.c_str());
        kv->Remove(key);
    }
} 

File: src/MemoryKVLib.Net.Tests/BoundaryTests.cs
using Xunit;
using MemoryKVLib.Net;
using System;
using System.Collections.Generic;
using System.Text;

namespace MemoryKVLib.Net.Tests
{
    public class BoundaryTests : IDisposable
    {
        private readonly MemoryKVClient _client;
        private const string ServiceName = "test_kv";

        public BoundaryTests()
        {
            _client = new MemoryKVClient();
            _client.Initialize(ServiceName);
        }

        public void Dispose()
        {
            _client.Dispose();
        }

        [Fact]
        public void MaxKeyLength()
        {
            // 测试最大长度键
            string maxKey = new string('a', 64);
            string value = "test_value";
            Assert.True(_client.Set(maxKey, value));
            
            Assert.True(_client.Get(maxKey, out string retrievedValue));
            Assert.Equal(value, retrievedValue);
            
            // 测试超长键
            string tooLongKey = new string('a', 65);
            Assert.False(_client.Set(tooLongKey, value));
        }

        [Fact]
        public void MaxValueLength()
        {
            // 测试最大长度值
            string key = "test_key";
            string maxValue = new string('a', 256);
            Assert.True(_client.Set(key, maxValue));
            
            Assert.True(_client.Get(key, out string retrievedValue));
            Assert.Equal(maxValue, retrievedValue);
            
            // 测试超长值
            string tooLongValue = new string('a', 257);
            Assert.False(_client.Set(key, tooLongValue));
        }

        [Fact]
        public void MaxConcurrentConnections()
        {
            const int maxConnections = 1000;
            var clients = new List<MemoryKVClient>();
            
            // 创建最大数量的连接
            for (int i = 0; i < maxConnections; i++)
            {
                var client = new MemoryKVClient();
                Assert.True(client.Initialize(ServiceName));
                clients.Add(client);
            }
            
            // 尝试创建额外连接
            using (var extraClient = new MemoryKVClient())
            {
                Assert.False(extraClient.Initialize(ServiceName));
            }
            
            // 清理
            foreach (var client in clients)
            {
                client.Dispose();
            }
        }

        [Fact]
        public void MemoryUsageLimit()
        {
            const int maxMemoryMB = 100; // 100MB 内存限制
            const int valueSize = 1024 * 1024; // 1MB 值大小
            
            // 尝试写入超过内存限制的数据
            string largeValue = new string('a', valueSize);
            for (int i = 0; i < maxMemoryMB + 1; i++)
            {
                string key = $"key_{i}";
                if (i < maxMemoryMB)
                {
                    Assert.True(_client.Set(key, largeValue));
                }
                else
                {
                    Assert.False(_client.Set(key, largeValue));
                }
            }
        }

        [Fact]
        public void EmptyKeyAndValue()
        {
            // 测试空键
            Assert.False(_client.Set("", "value"));
            
            // 测试空值
            Assert.True(_client.Set("key", ""));
            
            Assert.True(_client.Get("key", out string retrievedValue));
            Assert.Equal("", retrievedValue);
        }

        [Fact]
        public void SpecialCharacters()
        {
            // 测试包含特殊字符的键和值
            string specialKey = "key!@#$%^&*()";
            string specialValue = "value\n\t\r\0";
            
            Assert.True(_client.Set(specialKey, specialValue));
            
            Assert.True(_client.Get(specialKey, out string retrievedValue));
            Assert.Equal(specialValue, retrievedValue);
        }

        [Fact]
        public void BufferSize()
        {
            string key = "test_key";
            string value = "test_value";
            Assert.True(_client.Set(key, value));
            
            // 测试缓冲区不足
            Assert.False(_client.Get(key, out string retrievedValue, 5));
            
            // 测试缓冲区刚好
            Assert.True(_client.Get(key, out retrievedValue, value.Length + 1));
            Assert.Equal(value, retrievedValue);
        }
    }
} 